libtenzir/include/tenzir/import_helper.hpp

```diff
@@ -0,0 +1,124 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice.hpp>
+
+#include <caf/attach_stream_source.hpp>
+#include <caf/scheduled_actor.hpp>
+#include <caf/scoped_actor.hpp>
+#include <caf/typed_event_based_actor.hpp>
+
+namespace tenzir {
+
+namespace detail {
+
+class import_source_driver final
+  : public caf::stream_source_driver<
+      caf::broadcast_downstream_manager<table_slice>> {
+public:
+  import_source_driver(generator<table_slice>& input, uint64_t& num_events,
+                       std::function<void(caf::error)> on_finalize)
+    : input_{input},
+      num_events_{num_events},
+      on_finalize_{std::move(on_finalize)} {
+  }
+
+  void pull(caf::downstream<table_slice>& out, size_t num) override {
+    auto it = input_.unsafe_current();
+    for (size_t i = 0; i < num; ++i) {
+      TENZIR_ASSERT(it != input_.end());
+      ++it;
+      if (it == input_.end())
+        return;
+      auto next = std::move(*it);
+      if (next.rows() == 0) {
+        return;
+      }
+      num_events_ += next.rows();
+      next.import_time(time::clock::now());
+      out.push(std::move(next));
+    }
+  }
+
+  auto done() const noexcept -> bool override {
+    return input_.unsafe_current() == input_.end();
+  }
+
+  void finalize(const caf::error& error) override {
+    // TODO: Why do we catch `unreachable` here?
```

Comment:
You get an unreachable error from CAF when an actor handle's refcount drops to zero, so this is not really an error.